### PR TITLE
SQL GROUP BY RTS-732

### PR DESCRIPTION
**Pull request by andytill**
Syntax support for SQL GROUP BY and checks where the table DDL is not required.

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -12,18 +12,15 @@
...
                deprecated_function_calls, deprecated_functions]}.
 
-{eunit_opts, [
-verbose
-%     no_tty,  %% This turns off the default output, MUST HAVE
-%     {report, {eunit_progress, [colored, profile]}} %% Use `profile' to see test timing information
-     %% Uses the progress formatter with ANSI-colored output
-     ]}.
+
+{eunit_opts, [no_tty, {report, {unite_compact, []}}]}.
```
Comment (hazen) [inline rebar.config:16]: Maybe reinstate the comment about why we have `no_tty` set?  Also what does `unite_compact` specify?

File: rebar.config [hunk 2]
```diff
@@ -12,18 +12,15 @@
...
 
 %% == escriptize ==
 {escript_emu_args, "%%! -escript main riak_ql_cmd -smp disable +A 0\n"}.
 {escript_incl_apps, [sext]}.
 
 {deps, [
+        {unite, "", {git, "git://github.com/eproxus/unite.git", {tag, "v0.0.1"}}},
```
Comment (hazen) [inline rebar.config:23]: Who doesn't love pretty eunit tests?

File: src/riak_ql_parser.yrl [hunk 3]
```diff
@@ -377,7 +384,8 @@ Erlang code.
           fields  = [],
           limit   = [],
           where   = [],
-          ops     = []
+          ops     = [],
+          group_by
```
Comment (hazen) [inline src/riak_ql_parser.yrl:397]: Do we want to initialize to the empty list here, too?

File: src/riak_ql_parser.yrl [hunk 4]
```diff
@@ -416,30 +424,80 @@ convert(#outputs{type    = select,
...
+%% be specified in the GROUP BY.
+find_group_identifiers({identifier, [<<"*">>]} = Identifier, Acc) ->
+    [Identifier|Acc];
+find_group_identifiers({identifier, _} = Identifier, Acc) ->
+    [Identifier|Acc];
+find_group_identifiers({negate, Expr}, Acc) ->
+    find_group_identifiers(Expr, Acc);
+find_group_identifiers({Op, Left, Right}, Acc) when is_atom(Op) ->
+    find_group_identifiers(Right, find_group_identifiers(Left, Acc));
+find_group_identifiers({{window_agg_fn, _}, _}, Acc) ->
```
Comment (hazen) [inline src/riak_ql_parser.yrl:502]: We don't care which fields are being aggregated?

File: src/riak_ql_parser.yrl [hunk 5]
```diff
@@ -451,11 +509,12 @@ make_select({select, _SelectBytes},
...
-    _O = #outputs{type    = select,
-                  fields  = FieldsWrappedIdentifiers,
-                  buckets = Bucket,
-                  where   = E
-                 }.
+    #outputs{type    = select,
+             fields  = FieldsWrappedIdentifiers,
+             buckets = Bucket,
+             where   = E,
+             group_by = lists:flatten([GroupFields])
```
Comment (hazen) [inline src/riak_ql_parser.yrl:536]: I take it that each field is a list already

**Comment by hazen** [comment]
Unit tests all pass

File: rebar.config (hunk 2 above)
Comment (andytill) [inline rebar.config:23]: Crazy people.

File: src/riak_ql_parser.yrl (hunk 4 above)
Comment (andytill) [inline src/riak_ql_parser.yrl:502]: No, any field can be aggregated.

File: rebar.config (hunk 1 above)
Comment (andytill) [inline rebar.config:16]: `no_tty` is in the eunit docs though.

File: src/riak_ql_parser.yrl [hunk 6]
```diff
@@ -416,30 +424,80 @@ convert(#outputs{type    = select,
...
-%% make_atom({binary, SomeWord}) ->
-%%     {atom, binary_to_atom(SomeWord, utf8)}.
+validate_select_query(Outputs) ->
+    ok = assert_group_by_select(Outputs).
+
+%% If the query uses GROUP BY then check that the identifiers in the select
+%% all exist in the GROUP BY.
+assert_group_by_select(#outputs{ group_by = [] }) ->
+    ok;
+assert_group_by_select(#outputs{ fields = Fields, group_by = GroupBy }) ->
```
Comment (gordonguthrie) [inline src/riak_ql_parser.yrl]: `*` is a valid field identifier for GROUP BY's but this code throws a `Field(s) * are specified in the select statement but not the GROUP BY.` error

**Comment by gordonguthrie** [comment]
+1

**Comment by gordonguthrie** [comment]
+1 57ef47d

**Comment by andytill** [comment]
+1 094a898

**Comment by gordonguthrie** [comment]
+1 094a898

**Pull request merged**